File: client-android/app/src/androidTest/java/stasis/client_android/mocks/MockBackupTracker.kt
package stasis.client_android.mocks

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import stasis.client_android.lib.collection.rules.Rule
import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.model.SourceEntity
import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId
import stasis.client_android.lib.model.server.datasets.DatasetEntryId
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.tracking.BackupTracker
import stasis.client_android.lib.tracking.state.BackupState
import stasis.client_android.lib.utils.Either
import stasis.client_android.tracking.BackupTrackerManage
import stasis.client_android.tracking.BackupTrackerView
import java.nio.file.Path
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger

class MockBackupTracker : BackupTracker, BackupTrackerView, BackupTrackerManage {
    private val stats: Map<Statistic, AtomicInteger> = mapOf(
        Statistic.Started to AtomicInteger(0),
        Statistic.EntityDiscovered to AtomicInteger(0),
        Statistic.SpecificationProcessed to AtomicInteger(0),
        Statistic.EntityExamined to AtomicInteger(0),
        Statistic.EntitySkipped to AtomicInteger(0),
        Statistic.EntityCollected to AtomicInteger(0),
        Statistic.EntityProcessingStarted to AtomicInteger(0),
        Statistic.EntityPartProcessed to AtomicInteger(0),
        Statistic.EntityProcessed to AtomicInteger(0),
        Statistic.MetadataCollected to AtomicInteger(0),
        Statistic.MetadataPushed to AtomicInteger(0),
        Statistic.FailureEncountered to AtomicInteger(0),
        Statistic.Completed to AtomicInteger(0)
    )

    override val state: LiveData<Map<OperationId, BackupState>>
        get() = MutableLiveData(emptyMap())

    override fun updates(operation: OperationId): LiveData<BackupState> =
        MutableLiveData(BackupState.start(operation, UUID.randomUUID()))

    override fun remove(operation: OperationId) {}

    override fun clear() {}

    override suspend fun stateOf(operation: OperationId): BackupState? =
        null

    override fun started(operation: OperationId, definition: DatasetDefinitionId) {
        stats[Statistic.Started]?.getAndIncrement()
    }

    override fun entityDiscovered(
        operation: OperationId,
        entity: Path
    ) {
        stats[Statistic.EntityDiscovered]?.getAndIncrement()
    }

    override fun specificationProcessed(
        operation: OperationId,
        unmatched: List<Pair<Rule, Throwable>>
    ) {
        stats[Statistic.SpecificationProcessed]?.getAndIncrement()
    }

    override fun entityExamined(operation: OperationId, entity: Path) {
        stats[Statistic.EntityExamined]?.getAndIncrement()
    }

    override fun entitySkipped(operation: OperationId, entity: Path) {
        stats[Statistic.EntitySkipped]?.getAndIncrement()
    }

    override fun entityCollected(operation: OperationId, entity: SourceEntity) {
        stats[Statistic.EntityCollected]?.getAndIncrement()
    }

    override fun entityProcessingStarted(operation: OperationId, entity: Path, expectedParts: Int) {
        stats[Statistic.EntityProcessingStarted]?.getAndIncrement()
    }

    override fun entityPartProcessed(operation: OperationId, entity: Path) {
        stats[Statistic.EntityPartProcessed]?.getAndIncrement()
    }

    override fun entityProcessed(
        operation: OperationId,
        entity: Path,
        metadata: Either<EntityMetadata, EntityMetadata>
    ) {
        stats[Statistic.EntityProcessed]?.getAndIncrement()
    }

    override fun metadataCollected(operation: OperationId) {
        stats[Statistic.MetadataCollected]?.getAndIncrement()
    }

    override fun metadataPushed(operation: OperationId, entry: DatasetEntryId) {
        stats[Statistic.MetadataPushed]?.getAndIncrement()
    }

    override fun failureEncountered(operation: OperationId, entity: Path, failure: Throwable) {
        stats[Statistic.FailureEncountered]?.getAndIncrement()
    }

    override fun failureEncountered(operation: OperationId, failure: Throwable) {
        stats[Statistic.FailureEncountered]?.getAndIncrement()
    }

    override fun completed(operation: OperationId) {
        stats[Statistic.Completed]?.getAndIncrement()
    }

    val statistics: Map<Statistic, Int>
        get() = stats.mapValues { it.value.get() }

    sealed class Statistic {
        object Started : Statistic()
        object EntityDiscovered : Statistic()
        object SpecificationProcessed : Statistic()
        object EntitySkipped : Statistic()
        object EntityExamined : Statistic()
        object EntityCollected : Statistic()
        object EntityProcessingStarted : Statistic()
        object EntityPartProcessed : Statistic()
        object EntityProcessed : Statistic()
        object MetadataCollected : Statistic()
        object MetadataPushed : Statistic()
        object FailureEncountered : Statistic()
        object Completed : Statistic()
    }
}

File: client-ios/lib/Tests/StasisClientLibTests/Tracking/MockBackupTracker.swift
import Foundation
@testable import StasisClientLib

final class MockBackupTracker: BackupTracker {
    enum Statistic: String, CaseIterable, Sendable {
        case started = "Started"
        case entityDiscovered = "EntityDiscovered"
        case specificationProcessed = "SpecificationProcessed"
        case entityExamined = "EntityExamined"
        case entitySkipped = "EntitySkipped"
        case entityCollected = "EntityCollected"
        case entityProcessingStarted = "EntityProcessingStarted"
        case entityPartProcessed = "EntityPartProcessed"
        case entityProcessed = "EntityProcessed"
        case metadataCollected = "MetadataCollected"
        case metadataPushed = "MetadataPushed"
        case failureEncountered = "FailureEncountered"
        case completed = "Completed"
    }

    private let counter = StatsCounter<Statistic>()

    var statistics: [Statistic: Int] { counter.snapshot }

    func started(operation: OperationId, definition: DatasetDefinitionId) async { counter.increment(.started) }
    func entityDiscovered(operation: OperationId, entity: EntityRef) async { counter.increment(.entityDiscovered) }
    func specificationProcessed(operation: OperationId, unmatched: [(Rule, any Error)]) async {
        counter.increment(.specificationProcessed)
    }
    func entityExamined(operation: OperationId, entity: EntityRef) async { counter.increment(.entityExamined) }
    func entitySkipped(operation: OperationId, entity: EntityRef) async { counter.increment(.entitySkipped) }
    func entityCollected(operation: OperationId, entity: SourceEntity) async { counter.increment(.entityCollected) }
    func entityProcessingStarted(operation: OperationId, entity: EntityRef, expectedParts: Int) async {
        counter.increment(.entityProcessingStarted)
    }
    func entityPartProcessed(operation: OperationId, entity: EntityRef) async { counter.increment(.entityPartProcessed) }
    func entityProcessed(operation: OperationId, entity: EntityRef, metadata: Either<EntityMetadata, EntityMetadata>) async {
        counter.increment(.entityProcessed)
    }
    func metadataCollected(operation: OperationId) async { counter.increment(.metadataCollected) }
    func metadataPushed(operation: OperationId, entry: DatasetEntryId) async { counter.increment(.metadataPushed) }
    func failureEncountered(operation: OperationId, failure: any Error) async { counter.increment(.failureEncountered) }
    func failureEncountered(operation: OperationId, entity: EntityRef, failure: any Error) async {
        counter.increment(.failureEncountered)
    }
    func completed(operation: OperationId) async { counter.increment(.completed) }
    func stateOf(operation: OperationId) async -> BackupState? { nil }
}
